### ISSUE TITLE
as_uint() can not be called on expressions bigger than 64 bits

### ISSUE THREAD

**Issue by Fare9** [trailofbits/maat#123]
I've found while doing the next challenge: [ollvm](https://twitter.com/mr_phrazer/status/1533105691954712576) that Maat crashes while calculating the expression from the input.

The problem comes with an instruction like the next one:

```
0042c97a 4c 0f af c2        IMUL            R8,RDX
                                              $U36e80:16 = INT_SEXT R8
                                              $U36f00:16 = INT_SEXT RDX
                                              $U37000:16 = INT_MULT $U36e80:16, $U36f00:16
                                              R8 = INT_MULT R8, RDX
                                              $U37180:8 = SUBPIECE $U37000:16, 8:4
                                              $U13e00:16 = INT_SEXT R8
                                              CF = INT_NOTEQUAL $U13e00:16, $U37000:16
                                              OF = COPY CF
```

When one of the *INT_SEXT* instructions create a value of 128 bits, this value probably is used later in *INT_MULT* and the function **as_uint** is called, but internally **as_uint** checks the *size* value, and this value is in this case 128 bits, for what I see the **as_uint** function calls **cst_sign_trunc** that returns a *cst_t* type which is a *int64_t*, for that reason I suppose 128 bits are not supported in this case. Is it possible to truncate the size in the *INT_SEXT* and *INT_ZEXT* instructions when the maximum size of the register extended is lower than the output of the extension? I think that would solve the problem in this case, and as for now only x86-64 is supported, it would be possible to check if size is greater than 64.

I can provide code or other information if needed.

Thanks.

**Comment by Boyan-MILANOV** [comment]
Hey @Fare9, thanks for reporting this issue. Normally for values on more than `64` bits Maat will use `as_number()` which uses the `Number` class for arbitrary sized arithmetic. We must be calling `as_uint()` erroneously from some place... 

Can you provide me with a stack trace after Maat throws the exception? Or if its easier send me the script and the binary so I can reproduce the issue locally. 

Thanks!

**Comment by Fare9** [comment]
Yes I could send both, is there any email where I could send a zip file with both the CTF challenge and my script?

**Comment by Fare9** [comment]
Done! Sent both the binary and the script, I'll keep reading for news about it, I find interesting digging into Maat's source code.
Thanks.

**Comment by Fare9** [comment]
For what I see, the error happens in the next function:

```cpp
bool ite_evaluate(Expr l, ITECond cond, Expr r, const VarContext* ctx)
{
    if (ctx)
    {
        switch (cond)
        {
            case ITECond::EQ: return l->as_uint(*ctx) == r->as_uint(*ctx);
            case ITECond::LT: return l->as_uint(*ctx) < r->as_uint(*ctx);
            case ITECond::LE: return l->as_uint(*ctx) <= r->as_uint(*ctx);
            case ITECond::FEQ: return l->as_float(*ctx) == r->as_float(*ctx);
            case ITECond::FLE: return l->as_float(*ctx) <= r->as_float(*ctx);
            case ITECond::FLT: return l->as_float(*ctx) < r->as_float(*ctx);
            case ITECond::SLT: return l->as_int(*ctx) < r->as_int(*ctx);
            case ITECond::SLE: return l->as_int(*ctx) <= r->as_int(*ctx);
            default: break;
        }
    }
    else
    {
        switch (cond)
        {
            case ITECond::EQ: return l->as_uint() == r->as_uint();
            case ITECond::LT: return l->as_uint() < r->as_uint();
            case ITECond::LE: return l->as_uint() <= r->as_uint();
            case ITECond::FEQ: return l->as_float() == r->as_float();
            case ITECond::FLE: return l->as_float() <= r->as_float();
            case ITECond::FLT: return l->as_float() < r->as_float();
            case ITECond::SLT: return l->as_int() < r->as_int();
            case ITECond::SLE: return l->as_int() <= r->as_int();
            default: break;
        }
    }
    throw runtime_exception("ite_evaluate(): got unknown ITECond");
}
```

When it tries to run the next line:

```cpp
case ITECond::EQ: return l->as_uint() == r->as_uint();
```

Once it tries to get the **as_uint**, of any of those operands, it crashes because both operands are 128 bit. Maybe it would be possible to skip the evaluation of the constants or use the **as_number** as you said before, in that case the **Number** class would be compared.

**Comment by Fare9** [comment]
The next changes applied to my local copy make that the program doesn't crash:

```cpp
bool ite_evaluate(Expr l, ITECond cond, Expr r, const VarContext* ctx)
{
    if (ctx)
    {
        switch (cond)
        {
            case ITECond::EQ: return l->as_number(*ctx).get_ucst() == r->as_number(*ctx).get_ucst();
            case ITECond::LT: return l->as_number(*ctx).get_ucst() < r->as_number(*ctx).get_ucst();
            case ITECond::LE: return l->as_number(*ctx).get_ucst() <= r->as_number(*ctx).get_ucst();
            case ITECond::FEQ: return l->as_float(*ctx) == r->as_float(*ctx);
            case ITECond::FLE: return l->as_float(*ctx) <= r->as_float(*ctx);
            case ITECond::FLT: return l->as_float(*ctx) < r->as_float(*ctx);
            case ITECond::SLT: return l->as_number(*ctx).get_cst() < r->as_number(*ctx).get_cst();
            case ITECond::SLE: return l->as_number(*ctx).get_cst() <= r->as_number(*ctx).get_cst();
            default: break;
        }
    }
    else
    {
        switch (cond)
        {
            case ITECond::EQ: return l->as_number().get_ucst() == r->as_number().get_ucst();
            case ITECond::LT: return l->as_number().get_ucst() < r->as_number().get_ucst();
            case ITECond::LE: return l->as_number().get_ucst() <= r->as_number().get_ucst();
            case ITECond::FEQ: return l->as_float() == r->as_float();
            case ITECond::FLE: return l->as_float() <= r->as_float();
            case ITECond::FLT: return l->as_float() < r->as_float();
            case ITECond::SLT: return l->as_number().get_cst() < r->as_number().get_cst();
            case ITECond::SLE: return l->as_number().get_cst() <= r->as_number().get_cst();
            default: break;
        }
    }
    throw runtime_exception("ite_evaluate(): got unknown ITECond");
}
```

But I should also check the simplifications because it takes very loooooooong time to print the **Expression**, probably some simplification more would be necessary or clean std::ostream buffer after printing constants, but that would be another issue I think.

**Comment by Boyan-MILANOV** [comment]
Nice catch. Actually I have already fixed this in the `dev-evm` branch: https://github.com/trailofbits/maat/blob/4c8fad18b009d089a41ddb77c399e43b1af8f148/src/expression/expression.cpp#L2199-L2231

I am planning to merge `dev-evm` in `master` somewhere around next week, that should fix the issue :)

As regards printing big symbolic expressions... there isn't really a way to optimise it without truncating expressions. If the expression is huge it'll take a long time to print :/

**Comment by Fare9** [comment]
Yes, I saw it is also a problem in **Triton** if you call **unroll**,  so if the issue is already fixed in another branch probably this issue can be closed, thank you very much for the quick response. I'm trying to think why when trying to solve the constraint from the challenge, program keeps like in an idle state, if I find something regarding Z3 with Maat, I will write the issue.

Thanks!

**Comment by Boyan-MILANOV** [comment]
Fixed by #129 / #124
